charmhelpers/fetch/ubuntu.py

```diff
@@ -461,11 +566,62 @@ def _add_apt_repository(spec):
     """Add the spec using add_apt_repository
 
     :param spec: the parameter to pass to add_apt_repository
+    :type spec: str
     """
     if '{series}' in spec:
-        series = lsb_release()['DISTRIB_CODENAME']
+        series = get_distrib_codename()
         spec = spec.replace('{series}', series)
-    _run_with_retries(['add-apt-repository', '--yes', spec])
+    # software-properties package for bionic properly reacts to proxy settings
+    # passed as environment variables (See lp:1433761). This is not the case
+    # LTS and non-LTS releases below bionic.
+    _run_with_retries(['add-apt-repository', '--yes', spec],
+                      cmd_env=_env_proxy_settings(['https']))
+
+
+def _env_proxy_settings(selected_settings=None):
+    """Get proxy settings from process environment variables.
+
+    Get charm proxy settings from environment variables that correspond to
+    juju-http-proxy, juju-https-proxy and juju-no-proxy (available as of 2.4.2,
+    see lp:1782236) in a format suitable for passing to an application that
+    reacts to proxy settings passed as environment variables. Some applications
+    support lowercase or uppercase notation (e.g. curl), some support only
+    lowercase (e.g. wget), there are also subjectively rare cases of only
+    uppercase notation support. no_proxy CIDR and wildcard support also varies
+    between runtimes and applications as there is no enforced standard.
+
+    Some applications may connect to multiple destinations and expose config
+    options that would affect only proxy settings for a specific destination
+    these should be handled in charms in an application-specific manner.
+
+    :param selected_settings: format only a subset of possible settings
+    :type selected_settings: dict
+    :rtype: Option(None, dict[str, str])
+    """
+    SUPPORTED_SETTINGS = {
+        'http': 'HTTP_PROXY',
+        'https': 'HTTPS_PROXY',
+        'no_proxy': 'NO_PROXY',
```

Comment:
juju-no-proxy can contain subnets, which the NO_PROXY environment variable does not support. Are we introducing a bug by passing the value of juju-no-proxy to NO_PROXY without attempting to expand those subnets?

Reply:
@petevg 

I would rather leave it as-is because the underlying software may or may not support this notation.

add-apt-repository specifically doesn't need no_proxy because it (at worst) connects to two endpoints:

1. launchpad.net to resolve a ppa shortcut to a full deb line;
2. keyserver.ubuntu.com to download a GPG key for a repo to add.

Both of these will be accessed through a proxy in real-life scenarios. If they really need to be accessed without a proxy when proxy settings are configured they can be added to no_proxy as hostnames.

So there is no harm in the fact that curl invoked by add-apt-repository will ignore no_proxy.

The strong argument against expanding no_proxy in every situation is that expanded no_proxy can reach the limits imposed on argument size in terms of a child process stack memory - the decision on whether to expand no_proxy or not should be done by the caller of `_env_proxy_settings`:

https://elixir.bootlin.com/linux/v4.14.96/source/fs/exec.c#L244
```
static struct page *get_arg_page(struct linux_binprm *bprm, unsigned long pos,
		int write)
// ...
		/*
		 * We've historically supported up to 32 pages (ARG_MAX)
		 * of argument strings even with small stacks
		 */
		if (size <= ARG_MAX)
			return page;
// ...
		/*
		 * Limit to 1/4 of the max stack size or 3/4 of _STK_LIM
		 * (whichever is smaller) for the argv+env strings.
		 * This ensures that:
		 *  - the remaining binfmt code will not run out of stack space,
		 *  - the program will have a reasonable amount of stack left
		 *    to work from.
		 */
		limit = _STK_LIM / 4 * 3;
		limit = min(limit, rlimit(RLIMIT_STACK) / 4);
		if (size > limit)
			goto fail;

```
https://elixir.bootlin.com/linux/v4.14.96/source/include/uapi/linux/resource.h#L66
`#define _STK_LIM	(8*1024*1024)`

Reply:
@dshcherb Hmmm ... I'm tempted to say that we shouldn't include `NO_PROXY` in the default list of supported settings.

I guess the question is this: if I naively pass the result of this routine to _env_proxy_settings, without specifying any arguments, what is the best way for this to fail?

1) It ignores NO_PROXY by default, which means that I struggle to figure out why it isn't respecting my juju-no-proxy setting.
2) I have something in juju-no-proxy that will cause mischief, and my charm falls over as a result.

Reply:
@dshcherb After talking this over with the team, I think that it would be better to handle both the cases where no_proxy has a cidr, and where it is too big inline here.

If juju-no-proxy contains a cidr, attempt to expand it.
If the cidr is now too big, either log a warning, or throw an Exception.

The goal is to make it easy to troubleshoot if something does go wrong. And also to attempt to do the "right" thing in the first place.